### PR TITLE
eth: make debug_StorageRangeAt take a block hash or number

### PR DESCRIPTION
**Pull request by jwasinger**

### CODE REVIEW

File: eth/api.go [hunk 1]
```diff
@@ -409,11 +409,18 @@ type storageEntry struct {
...
+		block = api.eth.blockchain.GetBlockByNumber(uint64(number))
+	} else if hash, ok := blockNrOrHash.Hash(); ok {
+		block = api.eth.blockchain.GetBlockByHash(hash)
+	} else {
+		return StorageRangeResult{}, errors.New("either block number or block hash must be specified")
+	}
+
 	if block == nil {
-		return StorageRangeResult{}, fmt.Errorf("block %#x not found", blockHash)
+		return StorageRangeResult{}, fmt.Errorf("block %d not found", block)
```
Comment (holiman) [inline eth/api.go:423]: You cannot `%d` a `*types.Block`. And that `*types.Block` is also `nil` here. Maybe you meant to do something like this? 

```suggestion
		return StorageRangeResult{}, fmt.Errorf("block %v not found", blockNrOrHash)
```

File: eth/api.go [hunk 2]
```diff
@@ -409,11 +409,18 @@ type storageEntry struct {
 }
 
 // StorageRangeAt returns the storage at the given block height and transaction index.
-func (api *DebugAPI) StorageRangeAt(ctx context.Context, blockHash common.Hash, txIndex int, contractAddress common.Address, keyStart hexutil.Bytes, maxResult int) (StorageRangeResult, error) {
-	// Retrieve the block
-	block := api.eth.blockchain.GetBlockByHash(blockHash)
+func (api *DebugAPI) StorageRangeAt(ctx context.Context, blockNrOrHash rpc.BlockNumberOrHash, txIndex int, contractAddress common.Address, keyStart hexutil.Bytes, maxResult int) (StorageRangeResult, error) {
+	var block *types.Block
+	if number, ok := blockNrOrHash.Number(); ok {
```
Comment (s1na) [inline eth/api.go:414]: Or use `api.eth.APIBackend.BlockByNumberOrHash`. I'm not sure why we don't use backend so much in this file and directly use blockchain 🤔

Comment (holiman) [reply eth/api.go:414]: yeah that looks like a good alternative

Comment (jwasinger) [reply eth/api.go]: I moved `APIBackend.BlockByNumberOrHash/BlockByNumber` into the `Ethereum` object.

Comment (jwasinger) [reply eth/api.go]: Oops.  that was uneccesary.  `Ethereum` has a pointer to the `APIBackend`...

Comment (jwasinger) [reply eth/api.go]: Okay.  should be good now.

**Review by s1na — APPROVED** [review]
LGTM!

**Review by holiman — APPROVED** [review]
LGTM

**Comment by holiman** [comment]
Sorry, I bitrotted this PR a bit, needs to apply the change against the file `eth/api_debug.go` now instead

**Review by rjl493456442 — APPROVED** [review]
Looks good to me. Although I think it's a breaking change, originally users send hash `0x1234` now {"blockHash": "0x1234"} is required. But it's a debug API anyway, it's totally fine to change it.

**Review by MariusVanDerWijden — APPROVED** [review]

**Comment by s1na** [comment]
> Although I think it's a breaking change, originally users send hash 0x1234 now {"blockHash": "0x1234"} is required. But it's a debug API anyway, it's totally fine to change it.

No it's not a breaking change. `rpc.BlockNumberOrHash`'s unmarshaling rules are flexible. It accepts both `{blockNumber: '', blockHash: ''}` as well as `'0x1'` or `0xhash`. `debug.traceCall` takes this as parameter. That's what I tested with.

```
> debug.traceCall({}, '0x61276fed11a014a7e2b04957033815f385931098c8179668612eb138e4b09683')
{
  failed: false,
  gas: 53000,
  returnValue: "",
  structLogs: []
}
```

**Pull request merged**